File: rust/binette/src/registry.rs
use std::collections::{HashMap, HashSet};

use crate::error::SchemaError;
use crate::hash::{primitive_for_type_id, primitive_type_id, schema_type_id};
use crate::schema::{Schema, SchemaBundle, SchemaKind, TypeId, TypeRef, VariantPayload};

#[derive(Debug, Default, Clone)]
pub struct SchemaRegistry {
    schemas: HashMap<TypeId, Schema>,
}

impl SchemaRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn len(&self) -> usize {
        self.schemas.len()
    }

    pub fn is_empty(&self) -> bool {
        self.schemas.is_empty()
    }

    pub fn contains(&self, type_id: TypeId) -> bool {
        primitive_for_type_id(type_id).is_some() || self.schemas.contains_key(&type_id)
    }

    pub fn get(&self, type_id: TypeId) -> Option<&Schema> {
        self.schemas.get(&type_id)
    }

    pub fn schemas(&self) -> impl Iterator<Item = &Schema> {
        self.schemas.values()
    }

    // r[impl binette.schema.registry.install]
    // r[impl binette.bundle.registry]
    pub fn install_bundle(&mut self, bundle: &SchemaBundle) -> Result<(), SchemaError> {
        let batch = VerifiedSchemaBatch::new(&self.schemas, &bundle.schemas)?;
        batch.validate_type_ref(&bundle.root, &[])?;
        for attachment in &bundle.attachments {
            if let Some(metadata_schema) = &attachment.metadata_schema {
                batch.validate_type_ref(metadata_schema, &[])?;
            }
        }

        for schema in batch.into_schemas() {
            self.schemas.insert(schema.id, schema);
        }

        Ok(())
    }
}

struct VerifiedSchemaBatch<'a> {
    existing: &'a HashMap<TypeId, Schema>,
    schemas: HashMap<TypeId, Schema>,
}

impl<'a> VerifiedSchemaBatch<'a> {
    fn new(
        existing: &'a HashMap<TypeId, Schema>,
        incoming: &[Schema],
    ) -> Result<Self, SchemaError> {
        let mut schemas = HashMap::new();

        for schema in incoming {
            let computed = schema_type_id(schema)?;
            if computed != schema.id {
                return Err(SchemaError::SchemaIdMismatch {
                    declared: schema.id,
                    computed,
                });
            }

            if let SchemaKind::Primitive(primitive) = schema.kind
                && schema.id == primitive_type_id(primitive)
            {
                continue;
            }

            if let Some(primitive) = primitive_for_type_id(schema.id) {
                return Err(SchemaError::SchemaIdReservedForPrimitive {
                    type_id: schema.id,
                    primitive,
                });
            }

            if let Some(installed) = existing.get(&schema.id) {
                if installed != schema {
                    return Err(SchemaError::DuplicateSchemaId { type_id: schema.id });
                }
                continue;
            }

            if let Some(previous) = schemas.get(&schema.id) {
                if previous != schema {
                    return Err(SchemaError::DuplicateSchemaId { type_id: schema.id });
                }
                continue;
            }

            schemas.insert(schema.id, schema.clone());
        }

        let batch = Self { existing, schemas };
        batch.validate_schemas()?;
        batch.reject_batch_cycles()?;
        Ok(batch)
    }

    fn into_schemas(self) -> Vec<Schema> {
        self.schemas.into_values().collect()
    }

    fn validate_schemas(&self) -> Result<(), SchemaError> {
        for schema in self.schemas.values() {
            self.validate_type_params(&schema.type_params)?;
            self.validate_kind(&schema.kind, &schema.type_params)?;
        }
        Ok(())
    }

    fn validate_type_params(&self, type_params: &[String]) -> Result<(), SchemaError> {
        let mut seen = HashSet::new();
        for type_param in type_params {
            if !seen.insert(type_param) {
                return Err(SchemaError::DuplicateTypeParameter {
                    name: type_param.clone(),
                });
            }
        }
        Ok(())
    }

    fn validate_kind(&self, kind: &SchemaKind, scope: &[String]) -> Result<(), SchemaError> {
        match kind {
            SchemaKind::Primitive(_) | SchemaKind::Dynamic => Ok(()),
            SchemaKind::Struct { name, fields } => {
                self.validate_name(name)?;
                for field in fields {
                    self.validate_type_ref(&field.type_ref, scope)?;
                }
                Ok(())
            }
            SchemaKind::Enum { name, variants } => {
                self.validate_name(name)?;
                for variant in variants {
                    self.validate_variant_payload(&variant.payload, scope)?;
                }
                Ok(())
            }
            SchemaKind::Tuple { elements } => {
                if elements.is_empty() {
                    return Err(SchemaError::InvalidTupleArity);
                }
                for element in elements {
                    self.validate_type_ref(element, scope)?;
                }
                Ok(())
            }
            SchemaKind::List { element } | SchemaKind::Set { element } => {
                self.validate_type_ref(element, scope)
            }
            SchemaKind::Map { key, value } => {
                self.validate_type_ref(key, scope)?;
                self.validate_type_ref(value, scope)
            }
            SchemaKind::Array {
                element,
                dimensions,
            } => {
                if dimensions.is_empty() {
                    return Err(SchemaError::InvalidArrayRank);
                }
                self.validate_type_ref(element, scope)
            }
            SchemaKind::Option { element } => self.validate_type_ref(element, scope),
            // r[impl binette.schema.external]
            SchemaKind::External { kind, metadata: _ } => self.validate_name(kind),
        }
    }

    fn validate_variant_payload(
        &self,
        payload: &VariantPayload,
        scope: &[String],
    ) -> Result<(), SchemaError> {
        match payload {
            VariantPayload::Unit => Ok(()),
            VariantPayload::Newtype { type_ref } => self.validate_type_ref(type_ref, scope),
            VariantPayload::Tuple { elements } => {
                for element in elements {
                    self.validate_type_ref(element, scope)?;
                }
                Ok(())
            }
            VariantPayload::Struct { fields } => {
                for field in fields {
                    self.validate_type_ref(&field.type_ref, scope)?;
                }
                Ok(())
            }
        }
    }

    fn validate_name(&self, name: &str) -> Result<(), SchemaError> {
        if name.is_empty() {
            Err(SchemaError::EmptySchemaName)
        } else {
            Ok(())
        }
    }

    fn validate_type_ref(&self, type_ref: &TypeRef, scope: &[String]) -> Result<(), SchemaError> {
        match type_ref {
            TypeRef::Concrete { type_id, args } => {
                if primitive_for_type_id(*type_id).is_some() {
                    if !args.is_empty() {
                        return Err(SchemaError::TypeArgumentArity {
                            type_id: *type_id,
                            expected: 0,
                            actual: args.len(),
                        });
                    }
                } else {
                    let schema = self
                        .schema(*type_id)
                        .ok_or(SchemaError::UnknownTypeId { type_id: *type_id })?;
                    if schema.type_params.len() != args.len() {
                        return Err(SchemaError::TypeArgumentArity {
                            type_id: *type_id,
                            expected: schema.type_params.len(),
                            actual: args.len(),
                        });
                    }
                }

                for arg in args {
                    self.validate_type_ref(arg, scope)?;
                }

                Ok(())
            }
            TypeRef::Var { name } => {
                if scope.contains(name) {
                    Ok(())
                } else {
                    Err(SchemaError::UnknownTypeParameter { name: name.clone() })
                }
            }
        }
    }

    fn schema(&self, type_id: TypeId) -> Option<&Schema> {
        self.schemas
            .get(&type_id)
            .or_else(|| self.existing.get(&type_id))
    }

    fn reject_batch_cycles(&self) -> Result<(), SchemaError> {
        let mut visiting = HashSet::new();
        let mut visited = HashSet::new();

        for type_id in self.schemas.keys().copied() {
            self.visit_for_cycles(type_id, &mut visiting, &mut visited)?;
        }

        Ok(())
    }

    fn visit_for_cycles(
        &self,
        type_id: TypeId,
        visiting: &mut HashSet<TypeId>,
        visited: &mut HashSet<TypeId>,
    ) -> Result<(), SchemaError> {
        if !self.schemas.contains_key(&type_id) || visited.contains(&type_id) {
            return Ok(());
        }

        if !visiting.insert(type_id) {
            return Err(SchemaError::RecursiveRegistryUnsupported { type_id });
        }

        let mut deps = Vec::new();
        self.collect_batch_deps(
            &self
                .schemas
                .get(&type_id)
                .expect("cycle walk only starts from batch schemas")
                .kind,
            &mut deps,
        );

        for dep in deps {
            self.visit_for_cycles(dep, visiting, visited)?;
        }

        visiting.remove(&type_id);
        visited.insert(type_id);
        Ok(())
    }

    fn collect_batch_deps(&self, kind: &SchemaKind, out: &mut Vec<TypeId>) {
        match kind {
            SchemaKind::Primitive(_) | SchemaKind::Dynamic | SchemaKind::External { .. } => {}
            SchemaKind::Struct { fields, .. } => {
                for field in fields {
                    self.collect_type_ref_batch_deps(&field.type_ref, out);
                }
            }
            SchemaKind::Enum { variants, .. } => {
                for variant in variants {
                    self.collect_variant_batch_deps(&variant.payload, out);
                }
            }
            SchemaKind::Tuple { elements } => {
                for element in elements {
                    self.collect_type_ref_batch_deps(element, out);
                }
            }
            SchemaKind::List { element } | SchemaKind::Set { element } => {
                self.collect_type_ref_batch_deps(element, out);
            }
            SchemaKind::Map { key, value } => {
                self.collect_type_ref_batch_deps(key, out);
                self.collect_type_ref_batch_deps(value, out);
            }
            SchemaKind::Array { element, .. } | SchemaKind::Option { element } => {
                self.collect_type_ref_batch_deps(element, out);
            }
        }
    }

    fn collect_variant_batch_deps(&self, payload: &VariantPayload, out: &mut Vec<TypeId>) {
        match payload {
            VariantPayload::Unit => {}
            VariantPayload::Newtype { type_ref } => self.collect_type_ref_batch_deps(type_ref, out),
            VariantPayload::Tuple { elements } => {
                for element in elements {
                    self.collect_type_ref_batch_deps(element, out);
                }
            }
            VariantPayload::Struct { fields } => {
                for field in fields {
                    self.collect_type_ref_batch_deps(&field.type_ref, out);
                }
            }
        }
    }

    fn collect_type_ref_batch_deps(&self, type_ref: &TypeRef, out: &mut Vec<TypeId>) {
        match type_ref {
            TypeRef::Concrete { type_id, args } => {
                if self.schemas.contains_key(type_id) {
                    out.push(*type_id);
                }
                for arg in args {
                    self.collect_type_ref_batch_deps(arg, out);
                }
            }
            TypeRef::Var { .. } => {}
        }
    }
}

File: typescript/binette/src/registry.ts
import { BinetteError, encodeSelfDescribed } from "./value.js";

import {
  primitiveForTypeId,
  primitiveTypeId,
  recursiveTypeIdMap,
  schemaToValue,
  schemaTypeId,
  type Schema,
  type SchemaBundle,
  type SchemaKind,
  type TypeId,
  type TypeRef,
  type VariantPayload,
} from "./schema.js";

// r[impl binette.schema.registry+2]
export class SchemaRegistry {
  readonly #schemas = new Map<TypeId, Schema>();

  get size(): number {
    return this.#schemas.size;
  }

  get isEmpty(): boolean {
    return this.#schemas.size === 0;
  }

  contains(typeId: TypeId): boolean {
    return primitiveForTypeId(typeId) !== null || this.#schemas.has(typeId);
  }

  get(typeId: TypeId): Schema | undefined {
    return this.#schemas.get(typeId);
  }

  schemas(): IterableIterator<Schema> {
    return this.#schemas.values();
  }

  // r[impl binette.bundle.self-contained]
  validateSelfContainedBundle(bundle: SchemaBundle): void {
    VerifiedSchemaBatch.forSelfContained(this.#schemas, bundle);
  }

  // r[impl binette.schema.registry.install]
  // r[impl binette.bundle.registry]
  installBundle(bundle: SchemaBundle): void {
    const batch = VerifiedSchemaBatch.create(this.#schemas, bundle.schemas);
    batch.validateTypeRef(bundle.root, []);
    for (const attachment of bundle.attachments) {
      if (attachment.metadataSchema !== null) {
        batch.validateTypeRef(attachment.metadataSchema, []);
      }
    }

    for (const schema of batch.intoSchemas()) {
      this.#schemas.set(schema.id, schema);
    }
  }
}

class VerifiedSchemaBatch {
  readonly #existing: ReadonlyMap<TypeId, Schema>;
  readonly #schemas: Map<TypeId, Schema>;

  private constructor(
    existing: ReadonlyMap<TypeId, Schema>,
    schemas: Map<TypeId, Schema>,
  ) {
    this.#existing = existing;
    this.#schemas = schemas;
  }

  static create(
    existing: ReadonlyMap<TypeId, Schema>,
    incoming: readonly Schema[],
  ): VerifiedSchemaBatch {
    const batch = this.collect(existing, incoming);
    batch.validateSchemas();
    batch.verifyDeclaredIds();
    return batch;
  }

  static forSelfContained(
    existing: ReadonlyMap<TypeId, Schema>,
    bundle: SchemaBundle,
  ): VerifiedSchemaBatch {
    const batch = this.collect(existing, bundle.schemas);
    batch.validateSelfContainedBundle(bundle);
    batch.validateSchemas();
    batch.verifyDeclaredIds();
    return batch;
  }

  private static collect(
    existing: ReadonlyMap<TypeId, Schema>,
    incoming: readonly Schema[],
  ): VerifiedSchemaBatch {
    const schemas = new Map<TypeId, Schema>();

    for (const schema of incoming) {
      if (
        schema.kind.kind === "primitive" &&
        schema.id === primitiveTypeId(schema.kind.primitive)
      ) {
        continue;
      }

      const primitive = primitiveForTypeId(schema.id);
      if (primitive !== null) {
        throw new BinetteError(
          `schema id ${schema.id} is reserved for primitive ${primitive}`,
        );
      }

      const installed = existing.get(schema.id);
      if (installed !== undefined) {
        if (!schemasEqual(installed, schema)) {
          throw new BinetteError(`duplicate schema id ${schema.id}`);
        }
        continue;
      }

      const previous = schemas.get(schema.id);
      if (previous !== undefined) {
        if (!schemasEqual(previous, schema)) {
          throw new BinetteError(`duplicate schema id ${schema.id}`);
        }
        continue;
      }

      schemas.set(schema.id, schema);
    }

    return new VerifiedSchemaBatch(existing, schemas);
  }

  intoSchemas(): Schema[] {
    return Array.from(this.#schemas.values());
  }

  validateTypeRef(typeRef: TypeRef, scope: readonly string[]): void {
    switch (typeRef.kind) {
      case "concrete": {
        const primitive = primitiveForTypeId(typeRef.typeId);
        if (primitive !== null) {
          if (typeRef.args.length !== 0) {
            throw new BinetteError(
              `type id ${typeRef.typeId} expects 0 type arguments, got ${typeRef.args.length}`,
            );
          }
        } else {
          const schema = this.schema(typeRef.typeId);
          if (schema === undefined) {
            throw new BinetteError(`unknown type id ${typeRef.typeId}`);
          }
          if (schema.typeParams.length !== typeRef.args.length) {
            throw new BinetteError(
              `type id ${typeRef.typeId} expects ${schema.typeParams.length} type arguments, got ${typeRef.args.length}`,
            );
          }
        }

        for (const arg of typeRef.args) {
          this.validateTypeRef(arg, scope);
        }
        break;
      }
      case "var":
        if (!scope.includes(typeRef.name)) {
          throw new BinetteError(`unknown type parameter ${typeRef.name}`);
        }
        break;
    }
  }

  private validateSelfContainedBundle(bundle: SchemaBundle): void {
    const visited = new Set<TypeId>();
    this.validateSelfContainedTypeRef(bundle.root, [], visited);
    for (const attachment of bundle.attachments) {
      if (attachment.metadataSchema !== null) {
        this.validateSelfContainedTypeRef(
          attachment.metadataSchema,
          [],
          visited,
        );
      }
    }
  }

  private validateSchemas(): void {
    for (const schema of this.#schemas.values()) {
      this.validateTypeParams(schema.typeParams);
      this.validateKind(schema.kind, schema.typeParams);
    }
  }

  private validateTypeParams(typeParams: readonly string[]): void {
    const seen = new Set<string>();
    for (const typeParam of typeParams) {
      if (seen.has(typeParam)) {
        throw new BinetteError(`duplicate type parameter ${typeParam}`);
      }
      seen.add(typeParam);
    }
  }

  private validateKind(kind: SchemaKind, scope: readonly string[]): void {
    switch (kind.kind) {
      case "primitive":
      case "dynamic":
        break;
      case "struct":
        validateName(kind.name);
        for (const field of kind.fields) {
          this.validateTypeRef(field.typeRef, scope);
        }
        break;
      case "enum":
        validateName(kind.name);
        for (const variant of kind.variants) {
          this.validateVariantPayload(variant.payload, scope);
        }
        break;
      case "tuple":
        if (kind.elements.length === 0) {
          throw new BinetteError("tuple schema must not be empty");
        }
        for (const element of kind.elements) {
          this.validateTypeRef(element, scope);
        }
        break;
      case "list":
      case "set":
      case "option":
        this.validateTypeRef(kind.element, scope);
        break;
      case "map":
        this.validateTypeRef(kind.key, scope);
        this.validateTypeRef(kind.value, scope);
        break;
      case "array":
        if (kind.dimensions.length === 0) {
          throw new BinetteError("array schema dimensions must not be empty");
        }
        this.validateTypeRef(kind.element, scope);
        break;
      case "external":
        validateName(kind.externalKind);
        break;
    }
  }

  private validateVariantPayload(
    payload: VariantPayload,
    scope: readonly string[],
  ): void {
    switch (payload.kind) {
      case "unit":
        break;
      case "newtype":
        this.validateTypeRef(payload.typeRef, scope);
        break;
      case "tuple":
        for (const element of payload.elements) {
          this.validateTypeRef(element, scope);
        }
        break;
      case "struct":
        for (const field of payload.fields) {
          this.validateTypeRef(field.typeRef, scope);
        }
        break;
    }
  }

  private validateSelfContainedTypeRef(
    typeRef: TypeRef,
    scope: readonly string[],
    visited: Set<TypeId>,
  ): void {
    switch (typeRef.kind) {
      case "concrete": {
        const primitive = primitiveForTypeId(typeRef.typeId);
        if (primitive !== null) {
          if (typeRef.args.length !== 0) {
            throw new BinetteError(
              `type id ${typeRef.typeId} expects 0 type arguments, got ${typeRef.args.length}`,
            );
          }
        } else {
          const schema = this.schema(typeRef.typeId);
          if (schema === undefined) {
            throw new BinetteError(`missing bundle schema ${typeRef.typeId}`);
          }
          if (schema.typeParams.length !== typeRef.args.length) {
            throw new BinetteError(
              `type id ${typeRef.typeId} expects ${schema.typeParams.length} type arguments, got ${typeRef.args.length}`,
            );
          }
          if (!visited.has(typeRef.typeId)) {
            visited.add(typeRef.typeId);
            this.validateSelfContainedKind(schema.kind, schema.typeParams, visited);
          }
        }

        for (const arg of typeRef.args) {
          this.validateSelfContainedTypeRef(arg, scope, visited);
        }
        break;
      }
      case "var":
        if (!scope.includes(typeRef.name)) {
          throw new BinetteError(`unknown type parameter ${typeRef.name}`);
        }
        break;
    }
  }

  private validateSelfContainedKind(
    kind: SchemaKind,
    scope: readonly string[],
    visited: Set<TypeId>,
  ): void {
    switch (kind.kind) {
      case "primitive":
      case "dynamic":
      case "external":
        break;
      case "struct":
        for (const field of kind.fields) {
          this.validateSelfContainedTypeRef(field.typeRef, scope, visited);
        }
        break;
      case "enum":
        for (const variant of kind.variants) {
          this.validateSelfContainedVariantPayload(
            variant.payload,
            scope,
            visited,
          );
        }
        break;
      case "tuple":
        for (const element of kind.elements) {
          this.validateSelfContainedTypeRef(element, scope, visited);
        }
        break;
      case "list":
      case "set":
      case "option":
        this.validateSelfContainedTypeRef(kind.element, scope, visited);
        break;
      case "map":
        this.validateSelfContainedTypeRef(kind.key, scope, visited);
        this.validateSelfContainedTypeRef(kind.value, scope, visited);
        break;
      case "array":
        this.validateSelfContainedTypeRef(kind.element, scope, visited);
        break;
    }
  }

  private validateSelfContainedVariantPayload(
    payload: VariantPayload,
    scope: readonly string[],
    visited: Set<TypeId>,
  ): void {
    switch (payload.kind) {
      case "unit":
        break;
      case "newtype":
        this.validateSelfContainedTypeRef(payload.typeRef, scope, visited);
        break;
      case "tuple":
        for (const element of payload.elements) {
          this.validateSelfContainedTypeRef(element, scope, visited);
        }
        break;
      case "struct":
        for (const field of payload.fields) {
          this.validateSelfContainedTypeRef(field.typeRef, scope, visited);
        }
        break;
    }
  }

  // r[impl binette.schema.registry.recursive]
  private verifyDeclaredIds(): void {
    for (const component of this.stronglyConnectedComponents()) {
      const firstTypeId = component[0];
      if (firstTypeId === undefined) {
        throw new BinetteError("SCC is unexpectedly empty");
      }
      if (component.length === 1 && !this.hasSelfDependency(firstTypeId)) {
        const schema = mustGet(this.#schemas, firstTypeId);
        const computed = schemaTypeId(schema);
        if (computed !== schema.id) {
          throw new BinetteError(
            `schema declared id ${schema.id} but canonical content hashes to ${computed}`,
          );
        }
      } else {
        const schemas = component.map((typeId) => mustGet(this.#schemas, typeId));
        const computed = recursiveTypeIdMap(schemas);
        for (const schema of schemas) {
          const typeId = computed.get(schema.id);
          if (typeId === undefined) {
            throw new BinetteError(
              `recursive type-id map is missing schema ${schema.id}`,
            );
          }
          if (typeId !== schema.id) {
            throw new BinetteError(
              `schema declared id ${schema.id} but recursive content hashes to ${typeId}`,
            );
          }
        }
      }
    }
  }

  private stronglyConnectedComponents(): TypeId[][] {
    const state: TarjanState = {
      nextIndex: 0,
      indices: new Map(),
      lowlinks: new Map(),
      stack: [],
      onStack: new Set(),
      components: [],
    };

    for (const typeId of this.#schemas.keys()) {
      if (!state.indices.has(typeId)) {
        this.connectComponent(typeId, state);
      }
    }

    return state.components;
  }

  private connectComponent(typeId: TypeId, state: TarjanState): void {
    const index = state.nextIndex;
    state.nextIndex += 1;
    state.indices.set(typeId, index);
    state.lowlinks.set(typeId, index);
    state.stack.push(typeId);
    state.onStack.add(typeId);

    const deps: TypeId[] = [];
    this.collectBatchDeps(mustGet(this.#schemas, typeId).kind, deps);

    for (const dep of deps) {
      if (!state.indices.has(dep)) {
        this.connectComponent(dep, state);
        state.lowlinks.set(
          typeId,
          Math.min(mustGetNumber(state.lowlinks, typeId), mustGetNumber(state.lowlinks, dep)),
        );
      } else if (state.onStack.has(dep)) {
        state.lowlinks.set(
          typeId,
          Math.min(mustGetNumber(state.lowlinks, typeId), mustGetNumber(state.indices, dep)),
        );
      }
    }

    if (
      mustGetNumber(state.lowlinks, typeId) ===
      mustGetNumber(state.indices, typeId)
    ) {
      const component: TypeId[] = [];
      while (state.stack.length > 0) {
        const member = state.stack.pop();
        if (member === undefined) {
          throw new BinetteError("tarjan stack unexpectedly empty");
        }
        state.onStack.delete(member);
        component.push(member);
        if (member === typeId) {
          break;
        }
      }
      state.components.push(component);
    }
  }

  private hasSelfDependency(typeId: TypeId | undefined): boolean {
    if (typeId === undefined) {
      throw new BinetteError("SCC is unexpectedly empty");
    }
    const deps: TypeId[] = [];
    this.collectBatchDeps(mustGet(this.#schemas, typeId).kind, deps);
    return deps.includes(typeId);
  }

  private collectBatchDeps(kind: SchemaKind, out: TypeId[]): void {
    switch (kind.kind) {
      case "primitive":
      case "dynamic":
      case "external":
        break;
      case "struct":
        for (const field of kind.fields) {
          this.collectTypeRefBatchDeps(field.typeRef, out);
        }
        break;
      case "enum":
        for (const variant of kind.variants) {
          this.collectVariantBatchDeps(variant.payload, out);
        }
        break;
      case "tuple":
        for (const element of kind.elements) {
          this.collectTypeRefBatchDeps(element, out);
        }
        break;
      case "list":
      case "set":
      case "option":
        this.collectTypeRefBatchDeps(kind.element, out);
        break;
      case "map":
        this.collectTypeRefBatchDeps(kind.key, out);
        this.collectTypeRefBatchDeps(kind.value, out);
        break;
      case "array":
        this.collectTypeRefBatchDeps(kind.element, out);
        break;
    }
  }

  private collectVariantBatchDeps(payload: VariantPayload, out: TypeId[]): void {
    switch (payload.kind) {
      case "unit":
        break;
      case "newtype":
        this.collectTypeRefBatchDeps(payload.typeRef, out);
        break;
      case "tuple":
        for (const element of payload.elements) {
          this.collectTypeRefBatchDeps(element, out);
        }
        break;
      case "struct":
        for (const field of payload.fields) {
          this.collectTypeRefBatchDeps(field.typeRef, out);
        }
        break;
    }
  }

  private collectTypeRefBatchDeps(typeRef: TypeRef, out: TypeId[]): void {
    switch (typeRef.kind) {
      case "concrete":
        if (this.#schemas.has(typeRef.typeId)) {
          out.push(typeRef.typeId);
        }
        for (const arg of typeRef.args) {
          this.collectTypeRefBatchDeps(arg, out);
        }
        break;
      case "var":
        break;
    }
  }

  private schema(typeId: TypeId): Schema | undefined {
    return this.#schemas.get(typeId) ?? this.#existing.get(typeId);
  }
}

type TarjanState = {
  nextIndex: number;
  indices: Map<TypeId, number>;
  lowlinks: Map<TypeId, number>;
  stack: TypeId[];
  onStack: Set<TypeId>;
  components: TypeId[][];
};

function validateName(name: string): void {
  if (name.length === 0) {
    throw new BinetteError("schema names must not be empty");
  }
}

function schemasEqual(left: Schema, right: Schema): boolean {
  return bytesEqual(
    encodeSelfDescribed(schemaToValue(left)),
    encodeSelfDescribed(schemaToValue(right)),
  );
}

function bytesEqual(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }
  for (let index = 0; index < left.length; index += 1) {
    if (left[index] !== right[index]) {
      return false;
    }
  }
  return true;
}

function mustGet(map: ReadonlyMap<TypeId, Schema>, typeId: TypeId): Schema {
  const schema = map.get(typeId);
  if (schema === undefined) {
    throw new BinetteError(`missing batch schema ${typeId}`);
  }
  return schema;
}

function mustGetNumber(map: ReadonlyMap<TypeId, number>, typeId: TypeId): number {
  const value = map.get(typeId);
  if (value === undefined) {
    throw new BinetteError(`missing tarjan state for schema ${typeId}`);
  }
  return value;
}
